### PR TITLE
Publicize non-zero weierstrass affine variables

### PR DESCRIPTION
**Pull request by alex-ozdemir**
## Description

I've found these to be useful for a recent project.

I also added an `EqGadget` implementation and a test.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests 
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: src/groups/curves/short_weierstrass/mod.rs [hunk 1]
```diff
@@ -22,7 +22,9 @@ pub mod mnt4;
 ///  family of bilinear groups.
 pub mod mnt6;
 
-mod non_zero_affine;
+/// This module provides a generic implementation of elliptic curve operations for points on
+/// short-weierstrass curves in affine coordinates that **are not** equal to zero.
```
Comment (Pratyush) [inline src/groups/curves/short_weierstrass/mod.rs:26]: Can you add a warning stating that the `ProjectiveCurve` API should be the preferred one?

**Review by Pratyush — COMMENTED** [review]
I'm curious, what was your use case for the affine coordinates? Our projective coordinate impl is safe by default, but doesn't incur the overhead of the expensive complete formulae in scalar multiplication (by safely using on the incomplete affine formulae only when needed)

File: src/groups/curves/short_weierstrass/mod.rs (hunk 1 above)
Comment (alex-ozdemir) [reply src/groups/curves/short_weierstrass/mod.rs:26]: Yes! Done in 3ad9ca0.

**Comment by alex-ozdemir** [comment]
Sorry for the delay; I was out-of-town.

We were working with a prime-order curve and had a large computation that could produce intermediate zeros with only negligible probability.

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 2]
```diff
@@ -169,9 +169,65 @@ where
...
+        coordinates_equal.conditional_enforce_equal(&Boolean::Constant(true), condition)?;
+        Ok(())
+    }
+
+    #[inline]
+    #[tracing::instrument(target = "r1cs")]
+    fn enforce_equal(
+        &self,
+        other: &Self,
+    ) -> Result<(), SynthesisError> {
```
Comment (mmagician) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:207]:
```suggestion
    fn enforce_equal(&self, other: &Self) -> Result<(), SynthesisError> {
```

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 3]
```diff
@@ -300,4 +356,44 @@ mod test_non_zero_affine {
...
+                Ok(G1Parameters::AFFINE_GENERATOR_COEFFS.1)
+            })
+            .unwrap(),
+        );
+
+        let a = NonZeroAffineVar::<G1Parameters, FpVar<Fq>>::new(x, y);
+
+        let n = 10;
+
+        let a_multiples: Vec<NonZeroAffineVar::<G1Parameters, FpVar<Fq>>> =
```
Comment (mmagician) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:381]:
```suggestion
        let a_multiples: Vec<NonZeroAffineVar<G1Parameters, FpVar<Fq>>> =
```

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 4]
```diff
@@ -300,4 +356,44 @@ mod test_non_zero_affine {
...
+            std::iter::successors(Some(a.clone()), |acc| Some(acc.add_unchecked(&a).unwrap()))
+            .take(n)
+            .collect();
```
Comment (mmagician) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:384]:
```suggestion
                .take(n)
                .collect();
```

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 5]
```diff
@@ -300,4 +356,44 @@ mod test_non_zero_affine {
...
+
+        let all_equal: Vec<NonZeroAffineVar::<G1Parameters, FpVar<Fq>>> = (0..n/2)
+            .map(|i| a_multiples[i].add_unchecked(&a_multiples[n-i-1]).unwrap())
```
Comment (mmagician) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:387]:
```suggestion
        let all_equal: Vec<NonZeroAffineVar<G1Parameters, FpVar<Fq>>> = (0..n / 2)
            .map(|i| {
                a_multiples[i]
                    .add_unchecked(&a_multiples[n - i - 1])
                    .unwrap()
            )}
```

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 6]
```diff
@@ -300,4 +356,44 @@ mod test_non_zero_affine {
...
+            .collect();
+
+        for i in 0..n-1 {
+            a_multiples[i].enforce_not_equal(&a_multiples[i+1]).unwrap();
```
Comment (mmagician) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:391]:
```suggestion
        for i in 0..n - 1 {
            a_multiples[i]
                .enforce_not_equal(&a_multiples[i + 1])
                .unwrap();
```

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 7]
```diff
@@ -300,4 +356,44 @@ mod test_non_zero_affine {
...
+        }
+        for i in 0..all_equal.len()-1 {
+            all_equal[i].enforce_equal(&all_equal[i+1]).unwrap();
```
Comment (mmagician) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:394]:
```suggestion
        for i in 0..all_equal.len() - 1 {
            all_equal[i].enforce_equal(&all_equal[i + 1]).unwrap();
```

**Review by mmagician — CHANGES_REQUESTED** [review]

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 8]
```diff
@@ -300,4 +353,50 @@ mod test_non_zero_affine {
...
+            std::iter::successors(Some(a.clone()), |acc| Some(acc.add_unchecked(&a).unwrap()))
+                .take(n)
+                .collect();
+
+        let all_equal: Vec<NonZeroAffineVar<G1Parameters, FpVar<Fq>>> = (0..n / 2)
+            .map(|i| {
+                a_multiples[i]
+                    .add_unchecked(&a_multiples[n - i - 1])
+                    .unwrap()
+            )}
```
Comment (mmagician) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:388]:
```suggestion
            })
```

**Review by mmagician — APPROVED** [review]

**Pull request merged**